File: python/riot_sdk/facade/map_client.py
from __future__ import annotations

from typing import TYPE_CHECKING

from riot_sdk.core.business_response import ensure_success, require_response
from riot_sdk.core.maps import Map
from riot_sdk.core.stations import Station

if TYPE_CHECKING:
    from riot_sdk.facade.session import RiotSession


class MapClient:
    """Thin imap Map facade (BC-MAP-001 / ADR-sdk-0006)."""

    def __init__(self, session: RiotSession) -> None:
        self._session = session

    async def list_maps(self) -> list[Map]:
        """GET /api/imap/v1/mapInfo/getALLMapInfoExcludeMapJson (BC-MAP-001 / ADR-sdk-0005)."""
        client = self._session.create_generated_imap_client()
        response = require_response(
            await client.api.imap.v1.map_info.get_a_l_l_map_info_exclude_map_json.get(),
            "getALLMapInfoExcludeMapJson",
        )
        ensure_success(response.code, response.message)
        result = response.result or []
        return [
            Map(map_id=m.id, name=m.name)
            for m in result
            if m.id is not None and m.name
        ]

    async def list_stations(self, map_id: int) -> list[Station]:
        """GET /api/imap/v1/mapInfo/stations/{mapId} (BC-MAP-002 / ADR-sdk-0005)."""
        client = self._session.create_generated_imap_client()
        response = require_response(
            await client.api.imap.v1.map_info.stations.by_map_id(map_id).get(),
            f"stations/{map_id}",
        )
        ensure_success(response.code, response.message)
        result = response.result or []
        return [
            Station(map_id=map_id, station_id=s.id, name=s.name)
            for s in result
            if s.id is not None and s.name
        ]

    @property
    def raw(self):
        """Underlying Kiota imap client for endpoints not yet wrapped."""
        return self._session.create_generated_imap_client()

File: csharp/RIoT.Sdk.Facade/MapClient.cs
using RIoT.Sdk.Core;

namespace RIoT.Sdk.Facade;

/// <summary>
/// Thin imap Map facade (BC-MAP-001 / ADR-0007).
/// </summary>
public sealed class MapClient
{
    private readonly RiotSession _session;

    internal MapClient(RiotSession session) => _session = session;

    /// <summary>
    /// GET /api/imap/v1/mapInfo/getALLMapInfoExcludeMapJson — list Maps without mapJson (BC-MAP-001).
    /// Throws <see cref="RiotApiException"/> on business failure (ADR-0006).
    /// </summary>
    public async Task<IReadOnlyList<Map>> ListMapsAsync(CancellationToken cancellationToken = default)
    {
        var client = _session.CreateGeneratedImapClient();
        var response = await client.Api.Imap.V1.MapInfo.GetALLMapInfoExcludeMapJson
            .GetAsync(cancellationToken: cancellationToken)
            .ConfigureAwait(false);

        if (response is null)
        {
            throw new RiotApiException("getALLMapInfoExcludeMapJson returned empty response.");
        }

        if (!IsSuccessCode(response.Code))
        {
            throw new RiotApiException(
                $"RIoT business failure code={response.Code} message={response.Message}",
                statusCode: 200,
                businessCode: response.Code);
        }

        var result = response.Result ?? [];
        return result
            .Where(m => m.Id is not null && !string.IsNullOrWhiteSpace(m.Name))
            .Select(m => new Map(m.Id!.Value, m.Name!))
            .ToList();
    }

    /// <summary>
    /// GET /api/imap/v1/mapInfo/stations/{mapId} — list Stations on a Map (BC-MAP-002).
    /// Throws <see cref="RiotApiException"/> on business failure (ADR-0006).
    /// Empty list on success is a valid domain result (e.g. invalid mapId=0).
    /// </summary>
    public async Task<IReadOnlyList<Station>> ListStationsAsync(
        int mapId,
        CancellationToken cancellationToken = default)
    {
        var client = _session.CreateGeneratedImapClient();
        var response = await client.Api.Imap.V1.MapInfo.Stations[mapId]
            .GetAsync(cancellationToken: cancellationToken)
            .ConfigureAwait(false);

        if (response is null)
        {
            throw new RiotApiException($"stations/{mapId} returned empty response.");
        }

        if (!IsSuccessCode(response.Code))
        {
            throw new RiotApiException(
                $"RIoT business failure code={response.Code} message={response.Message}",
                statusCode: 200,
                businessCode: response.Code);
        }

        var result = response.Result ?? [];
        return result
            .Where(s => s.Id is not null && !string.IsNullOrWhiteSpace(s.Name))
            .Select(s => new Station(mapId, s.Id!.Value, s.Name!))
            .ToList();
    }

    /// <summary>Underlying Kiota imap client for endpoints not yet wrapped.</summary>
    public RIoT.Sdk.Generated.Imap.ImapClient Raw => _session.CreateGeneratedImapClient();

    private static bool IsSuccessCode(string? code)
        => string.IsNullOrWhiteSpace(code)
           || code is "0" or "200" or "OK" or "ok" or "success" or "SUCCESS";
}
